### ISSUE TITLE
C API (ERSA): build fails, undeclared type

### ISSUE THREAD

**Issue by didierersa** [openETCS/modeling#896]
The code ( from Green openETCS Non-Vital Demonstrator/Source Code/Source Code ETCS Onboard Unit System/Generated_Code/KCG-ERSA/) doesn't compile, I got this error:
...
Manage_TrackSideInformati_Manage_TrackSideInformation_Integration_Pkg.c:123:7: error: '_40_NC_CDDIFF_Specific_SSP_applicable_to_Cant_Deficiency_80_mm' undeclared (first use in this function)

...

**Comment by stefan-karg** [comment]
Dear Didier,

can you please send me the Makefile you are using to build the code?

Best regards
Stefan

**Comment by BerndHekele** [comment]
This comment is not clear. In order to get a proper understanding of the problems we need to be able to run this check in our environment and use the same settings of the C-Development. 
Do you have a project defined and is the  workbench useable on e.g., your demonstrator PC (the one we have in Munich)?

**Comment by didierersa** [comment]
In fact I use QMake as a makefile generator and I compile the whole scade source together our codeglue. Previous generated code (from release) was compiling fine. The problem arise when compiling MapLevelToColumn_InformationFilter_Pkg.c : 

gcc -c -pipe -fvisibility=hidden -g -std=gnu99 -O0 -ggdb3 -fPIC -D_REENTRANT -DTARGET="\"scade_debug\"" -DVERSION="\"1.0.0\"" -DORGANISATION_NAME="\"openETCS\"" -DORGANISATION_DOMAIN="\"openetcs.org\"" -DBUILD=1 -DQT_USE_FAST_CONCATENATION -DQT_USE_FAST_OPERATOR_PLUS -DQT_USE_QSTRINGBUILDER -D_REENTRANT=1 -D_ISOC99_SOURCE=1 -D_POSIX_SOURCE=1 -D_POSIX_C_SOURCE=200809L -D_GNU_SOURCE=1 -D__DEBUG__ -DQT_GUI_LIB -DQT_CORE_LIB -I../../../testrunner.git/evcglue/scade -I. -I../../../testrunner.git/evcglue/scade -isystem /usr/include/qt -isystem /usr/include/qt/QtGui -isystem /usr/include/qt/QtCore -I../../../testrunner.git/evcglue/socket94/.out/scade_debug/moc_debug -I/usr/lib/qt/mkspecs/linux-g++ -o ../../../testrunner.git/evcglue/socket94/.out/scade_debug/obj_debug/MapLevelToColumn_InformationFilter_Pkg.o ../../../testrunner.git/evcglue/scade/MapLevelToColumn_InformationFilter_Pkg.c


the _40_NC_CDDIFF_Specific_SSP_applicable_to_Cant_Deficiency_80_mm is never defined:

grep -rni "_40_NC_CDDIFF_Specific_SSP_applicable_to_Cant_Deficiency_80_mm" .
./evcglue/scade/Manage_TrackSideInformati_Manage_TrackSideInformation_Integration_Pkg.c:123:      _40_NC_CDDIFF_Specific_SSP_applicable_to_Cant_Deficiency_80_mm;
./evcglue/scade/Manage_TrackSideInformati_Manage_TrackSideInformation_Integration_Pkg.c:139:        _40_NC_CDDIFF_Specific_SSP_applicable_to_Cant_Deficiency_80_mm;

Something must not have been correctly exported

**Comment by didierersa** [comment]
I guess the correct value should be NC_CDDIFF_Specific_SSP_applicable_to_Cant_Deficiency_80_mm

**Comment by didierersa** [comment]
I've also undefined CompressedRadioMessage_TM and CompressedBaliseMessage_TM.

A simple grep -rni shows me that this symbols are defined in kcg_types.h ..from Green openETCS Non-Vital Demonstrator/Source Code/Source Code ETCS Track System/Generated_Code/AmsterdamUtrechtExpall/kcg_types.h

But unfortunately  not in the /kcg_types.h from Green openETCS Non-Vital Demonstrator/Source Code/Source Code ETCS Onboard Unit System/Generated_Code/KCG-ERSA

**Comment by JakobGartner** [comment]
it looks like you don’t have all the code  
we have to check

**Comment by stefan-karg** [comment]
Dear Didier,

I regenerated the code with the same settings (except for the global context) as the release code. Please check if the problems are solved now.

Stefan

**Comment by didierersa** [comment]
I will test ASAP, I will report back any issue I encounter.

**Comment by didierersa** [comment]
Dear stefan,

The issue is not new, but I've never reported it: some files, part of the scade installation are missing. This is not a big deal as I can take the one from my personal installation, but I wonder if something could happen if your scade version and mine doesn't match. I guess it would be safer if you put this files alongside with the generated code so mismatch won't happen:

libraries/libmathext/macro_libmathext.h
kcg_assign.h

**Comment by didierersa** [comment]
Unfortunately, I still got CompressedRadioMessage_TM and CompressedBaliseMessage_TM undefined. 

this symbols are defined in kcg_types.h ..from Green openETCS Non-Vital Demonstrator/Source Code/Source Code ETCS Track System/Generated_Code/AmsterdamUtrechtExpall/kcg_types.h

But unfortunately not in the kcg_types.h from Green openETCS Non-Vital Demonstrator/Source Code/Source Code ETCS Onboard Unit System/Generated_Code/KCG-ERSA

**Comment by didierersa** [comment]
To be precise, it is not the main scade model that do not compile it is the source of libenvsim. I don't know if I should compile it with or without scade support (I did with...)

**Comment by jokaICS** [comment]
You don't need to compile it -- I provide a static lib which you should be able to link in.

**Comment by jokaICS** [comment]
Here's the path to the lib:

https://github.com/openETCS/modeling/tree/master/test/EnvSim/libenvsim/dist/elf_x86

**Comment by didierersa** [comment]
Yes, I did noticed the static lib. Unfortunately, we choose to not use it because we want to be sure that our code can be at least compiled on any modern posix compliant platform without being stuck on a particular architecture (x86 / x86-64), or linux version or compiler. As source file were available, we just use them.

FYI, some time ago, a powerpc platform was a possibility and now I guess the GE platform is based on a ARM one. So being able to build from source is unfortunately not an option for us, especially for the glue code. After all it's an opensource project ;-)

So we will be very pleased if you can provide us a compilable version.

**Comment by didierersa** [comment]
And another missing c file, the one that contains definition of function:
../../../testrunner.git/evcglue/scade/CAST_DMI_TEXT_to_int_DATA_Variables.c:19: error: undefined reference to 'Char_to_Int_Utilities'
../../../testrunner.git/evcglue/scade/CAST_DriverIdentfier_to_int_DATA_Variables.c:19: error: undefined reference to 'Char_to_Int_Utilities'
../../../testrunner.git/evcglue/scade/CAST_int_to_DMI_TEXT_DATA_Variables.c:19: error: undefined reference to 'Int_to_Char_Utilities'
../../../testrunner.git/evcglue/scade/CAST_int_to_DriverIdentfier_DATA_Variables.c:19: error: undefined reference to 'Int_to_Char_Utilities'

**Comment by stefan-karg** [comment]
Dear Didier,

yesterday evening I've uploaded the missing c-files for the CAST operators mentioned above.

**Comment by didierersa** [comment]
Thank you very much

**Comment by stefan-karg** [comment]
Can we close the issue?

**Comment by didierersa** [comment]
It may not be related to this issue, but I still would like to have directly compilable source for the libenvsim, as sooner or later we will have to do it. For now we can start to test it with the static lib, but this is not optimal as some of our workstation are 64bits. Other issues can be considered closed.

**Comment by jokaICS** [comment]
The sources for libenvsim have always been available:

https://github.com/openETCS/modeling/tree/master/test/EnvSim/libenvsim

The `makex86.sh` located in that directory should get you started.

**Comment by didierersa** [comment]
I will try to compile it as soon as I've finished to integrate last work done by Jakob. Thanks for your work.
